native-sql-engine/core/src/main/scala/com/intel/oap/expression/ColumnarTernaryOperator.scala

```diff
@@ -56,14 +56,22 @@ class ColumnarSubString(str: Expression, pos: Expression, len: Expression, origi
     val (len_node, lenType): (TreeNode, ArrowType) =
       len.asInstanceOf[ColumnarExpression].doColumnarCodeGen(args)
 
-    //FIXME(): gandiva only support pos and len with int64 type
-    val lit_pos :ColumnarLiteral= pos.asInstanceOf[ColumnarLiteral]
-    val lit_pos_val = lit_pos.value
-    val long_pos_node = TreeBuilder.makeLiteral(lit_pos_val.asInstanceOf[Integer].longValue() :java.lang.Long)
-
-    val lit_len :ColumnarLiteral= len.asInstanceOf[ColumnarLiteral]
-    val lit_len_val = lit_len.value
-    val long_len_node = TreeBuilder.makeLiteral(lit_len_val.asInstanceOf[Integer].longValue() :java.lang.Long)
+    // FIXME(): gandiva only support pos and len with int64 type
+    val long_pos_node = pos match {
```

Comment:
Good work!
A small suggestion. It seems there is no need to tackle literal and non-literal case differently. I think just keeping line 64-65 is enough, which will cover both literal & non-literal cases. This should be also feasible to length node.

Reply:
This means we will add `castBIGINT` node even if it is a literal.

Reply:
Exactly. Please have a try.

Reply:
This works for me. I just conserned about the extra castBIGINT, which may cause more cpu costs.

Reply:
As discussed with Yuan, the overhead caused by cast may be ignorable.

Reply:
Okey, I will remove the `match` code.